corehq/motech/repeaters/expression/repeaters.py

```diff
@@ -145,23 +148,24 @@ def payload_doc(self, repeat_record):
 
     def allowed_to_forward(self, payload):
         allowed = super().allowed_to_forward(payload)
-        if allowed:
-            transactions = CaseTransaction.objects.get_last_n_recent_form_transaction(payload.case_id, 2)
-            # last 2 transactions were from repeater updates. This suggests a cycle.
-            possible_cycle = {t.xmlns for t in transactions} == {REPEATER_RESPONSE_XMLNS}
-            if possible_cycle:
-                logger.warning(
-                    f"Possible data forwarding loop detected for case {payload.case_id}. "
-                    f"Transactions: {[t.id for t in transactions]}"
-                )
+        if not allowed:
+            return False
+
+        transactions = CaseTransaction.objects.get_last_n_recent_form_transaction(
+            payload.case_id, MAX_REPEATER_CHAIN_LENGTH
+        )
+        for transaction in transactions:
+            if transaction.xmlns != REPEATER_RESPONSE_XMLNS:
+                # non-repeater update found, allow forwarding
+                return True
```

Comment:
Maybe esoteric, but would it be theoretically possible to circumvent the limit by having an endpoint update the case via API (for example) before returning a chained update response?

Likewise, any form submission that updates the case immediately breaks the chain limit detection, which is not much concern for "normal" usage, but seems vulnerable to timing attacks and possibly a concern for domains with very high submission throughput. Would it make sense to (maybe also) limit based on the number of transactions per time interval on a case?

Reply:
> Maybe esoteric, but would it be theoretically possible to circumvent the limit by having an endpoint update the case via API (for example) before returning a chained update response?

That is already possible now - there's not way to prevent external systems from using the APIs but we do have rate limiting applied to them as mentioned below.

> Would it make sense to (maybe also) limit based on the number of transactions per time interval on a case?

That seems like something that should be done at a central level since this is only one place that can auto-update cases. We do already have rate limiting applied to form submission processing (which would include these ones). Applying rate limiting to individual cases would also not fully solve the problem as you could have updates from one case trigger updates to another case and visa versa.

Overall I think the form submission rate limiting should be the main limiting factor for DDOS type issues. The checks here are mostly meant to prevent loops or excessive chaining which could rapidly consume the form submission rate limits for a project.